File: funcscript/funcs/text/EndsWithFunction.cs
using funcscript.core;
using System;
using funcscript.model;

namespace funcscript.funcs.strings
{
    internal class EndsWithFunction : IFsFunction, IFsDref
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Prefix;

        public string Symbol => "endswith";

        public int Precidence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != MaxParsCount)
                throw new error.TypeMismatchError($"{this.Symbol} function: Invalid parameter count. Expected {MaxParsCount}, but got {pars.Count}");

            var parBuilder = new CallRefBuilder(this,parent, pars);
            var par0 = parBuilder.GetParameter(0);
            var par1 = parBuilder.GetParameter(1);

            if (par0 is ValueReferenceDelegate || par1 is ValueReferenceDelegate)
                return parBuilder.CreateRef();

            return EvaluateInternal(par0, par1);
        }

        private object EvaluateInternal(object par0, object par1)
        {
            if (par0 == null || par1 == null)
                return false;

            if (!(par0 is string) || !(par1 is string))
                throw new error.TypeMismatchError($"Function {this.Symbol}. Both parameters must be strings");

            var mainString = (string)par0;
            var ending = (string)par1;

            return mainString.EndsWith(ending, StringComparison.Ordinal);
        }

        public object DrefEvaluate(IParameterList pars)
        {
            var par0 = FuncScript.Dref(pars.GetParameter(null, 0));
            var par1 = FuncScript.Dref(pars.GetParameter(null, 1));
            return EvaluateInternal(par0, par1);
        }

        public string ParName(int index)
        {
            return index switch
            {
                0 => "main string",
                1 => "ending substring",
                _ => null
            };
        }
    }
}

File: js-port/fs-js/src/funcs/text/ends-with-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');
const { requireString, makeValue, makeError, FsError, FSDataType } = helpers;

class EndsWithFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'endswith';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }
    const textResult = requireString(this.symbol, parameters.getParameter(provider, 0), 'text');
    if (!textResult.ok) {
      return textResult.error;
    }
    const endingResult = requireString(this.symbol, parameters.getParameter(provider, 1), 'ending substring');
    if (!endingResult.ok) {
      return endingResult.error;
    }
    return makeValue(FSDataType.Boolean, textResult.value.endsWith(endingResult.value));
  }
}

module.exports = {
  EndsWithFunction
};
